sponsorship_compassion/mappings/cancel_sponsorship_mapping.py

```diff
@@ -31,10 +30,69 @@ class CancelSponsorship(BaseSponsorshipMapping):
         "PrimaryHoldOwner": None
     }
 
+    params = {}
+
     def _process_connect_data(self, connect_data):
         # Set end date to correct format for Connect
-        if 'HoldExpirationDate' in connect_data:
-            endDateStr = connect_data.get('HoldExpirationDate')
-            endDate = datetime.strptime(endDateStr, "%Y-%m-%d %H:%M:%S")
-            connect_data['HoldExpirationDate'] = endDate.strftime(
+        if 'HoldExpirationDate' in connect_data and connect_data[
+                'HoldExpirationDate']:
+            end_date_str = connect_data['HoldExpirationDate']
+            end_date = datetime.strptime(end_date_str, "%Y-%m-%d %H:%M:%S")
+            connect_data['HoldExpirationDate'] = end_date.strftime(
                 "%Y-%m-%dT%H:%M:%SZ")
+
+            # save future hold with expiration date
+            child_global_id = connect_data['Beneficiary_GlobalID']
+            child = self.env['compassion.child'].search(
+                [('global_id', '=', child_global_id)], limit=1)
+
+            hold_vals = {
+                'name': "",
+                'child_id': child.id,
+                'type': 'Consignment Hold',
+                'expiration_date': end_date,
+                'primary_owner': 'Rose-Marie Reber',
+                'secondary_owner': 'Carole Rochat',
+                'no_money_yield_rate': '1.1',
+                'yield_rate': '1.1',
+                'channel': '',
+                'source_code': '',
+            }
+            self.env['compassion.hold'].create(hold_vals)
+
+    def _convert_connect_data(self, connect_name, value_mapping, value,
+                              relation_search=None):
+        if connect_name and connect_name in ['Beneficiary_GlobalID',
+                                             'HoldID']:
+            self.params[connect_name] = value
+
+        return super(CancelSponsorship, self)._convert_connect_data(
+            connect_name, value_mapping, value, relation_search=None)
+
+    def _process_odoo_data(self, connect_data):
+        if 'Beneficiary_GlobalID' in self.params:

```

Comment:
My proposition is to :

1. Before sending the message, when the wizard is executed to cancel the sponsorship, verify if the user wants to put a hold. If no, directly set the hold on the child (which is the hold that was set before the child is sponsored to inactive). If yes, keep the hold active.
2. Change base_mapping of field 'HoldId' to `'HoldID': 'hold_id'` (even if the field does not exist in recurring_contract), just that you can retrieve the value in function `_process_odoo_data`
3. remove `self.params` and change that test to 
```
if 'child_id' in connect_data:
    child = self.env['compassion.child'].browse(connect_data['child_id'])
    if child.hold_id.active:
        # Update new hold
        child.hold_id.hold_id = connect_data['hold_id']
        # Don't write hold_id in contract
        del connect_data['hold_id']
    else:
        child.active = False
```